llvm/docs/DirectX/DXILArchitecture.rst

```diff
@@ -0,0 +1,102 @@
+restrictions and various semantically important operations and
+metadata.
+
+LLVM's implementation philosophy for DXIL support is to treat DXIL as
+merely a representation format as much as possible. When reading DXIL,
+we should translate everyting to generic LLVM constructs when
+possible. Similarly, we should introduce DXIL-specific constructs as
+late as possible in the process of lowering to the format.
+
+There are three places to look for DXIL related code in LLVM: The
+`DirectX` backend, for writing DXIL; The `DXILUpgrade` pass, for
+reading; and in library code that is shared between writing and
+reading. We'll describe these in reverse order.
+
+Common Code for Reading and Writing
+===================================
+
+There's quite a bit of logic that needs to be shared between reading
+and writing DXIL in order to avoid code duplication. While we don't
+have a hard and fast rule about where such code should live, there are
+generally three sensible places. Simple definitions of enums and
+values that must stay fixed to match DXIL's ABI can be found in
+`Support/DXILABI.h`, utilities to translate bidirectionally between
+DXIL and modern LLVM constructs live in `lib/Transforms/Utils`, and
+more analyses that are needed to derive or preserve information are
+implemented as typical `lib/Analysis` passes.
+
+The DXILUpgrade Pass
+====================
+
+Translating DXIL to LLVM IR takes advantage of the fact that DXIL is
+compatible with LLVM 3.7 bitcode, and that modern LLVM is capable of
+"upgrading" older bitcode into modern IR. Simply relying on the
+bitcode upgrade process isn't sufficient though, since that leaves a
+number of DXIL specific constructs around. Thus, we have the
+`DXILUpgrade` pass to transform DXIL operations to LLVM operations and
+smooth over differences in metadata representation.
+
+The DirectX Backend
+===================
+
+The DirectX backend lowers LLVM IR into DXIL. As we're transforming to
+an intermediate format rather than a specific ISA, this backend does
+not follow the instruction selection patterns you might be familiar
+with from other backends. There are two parts to lowering DXIL - a set
+of passes that mutate various constructs into a form that matches how
+DXIL represents those constructs, followed by a limited bitcode
+"downgrader pass".
+
+Before emitting DXIL, the DirectX backend needs to modify the LLVM IR
+such that external operations, types, and metadata is represented in
+the way that DXIL expects. For example, `DXILOpLowering` translates
+intrinsics into `dx.op` calls. It's encouraged to implement parts of
+the DXIL lowering as these kinds of IR to IR passes when possible, as
+that means that they can be easily tested with `opt` and `FileCheck`
+without the need for external tooling.
+
+The second part of DXIL emission is more or less an LLVM bitcode
+downgrader. We need to emit bitcode that matches the LLVM 3.7
+representation. For this, we have `DXILWriter`, which is an alternate
```

Comment:
Just from a naming perspective it would be nice to have a DXILReader to match the DXILWriter, but it looks like we have DXILUpgrade instead?

Reply:
The nuance here is that we don't actually implement a DXILReader at all. As per the explanation of DXILUpgrade earlier in the doc:

> Translating DXIL to LLVM IR takes advantage of the fact that DXIL is
> compatible with LLVM 3.7 bitcode, and that modern LLVM is capable of
> "upgrading" older bitcode into modern IR. Simply relying on the
> bitcode upgrade process isn't sufficient though, ...

This is really the rationale for the somewhat unusual naming of DXILUpgrade, since it *isn't* the reader and it basically operates as an extension of LLVM's standard bitcode upgrade path. Let me know if there's some wording missing that could help clarify that.

Reply:
Yes, it makes sense that's what its doing, but that is also a bit of an implementation detail. I could also see us calling the Writer a "Downgrader" instead. In fact, that's what the doc says it does!

> The second part of DXIL emission is more or less an LLVM bitcode
downgrader.

It's not a big deal, but I think having a symmetric name for reader/writer or upgrader/downgrader would be nice to quickly understand that the two operations are inverses and how you would do a roundtrip.

Reply:
I'm a bit reticent to use a symmetric name here since they really aren't inverses on their own.
- The "downgrading" aspects of writing DXIL largely happen before we get to DXILWriter (which just writes the format), and it should indeed be possible to run the backend up to but not including DXILWriter, and then round trip that with DXILUpgrade.
- Similarly, DXILUpgrade does not read bitcode at all (let alone 3.7 era bitcode) - the standard LLVM bitcode reading/upgrade path gives us IR that's suitable input for DXILUpgrade, which then goes ahead and upgrades the DXIL specific constructs.

Reply:
I added sentences to the DXILUpgrade section and to the paragraph about lowering/downgrading to try to clarify this